### PR TITLE
frank/usermap

### PR DESCRIPTION
**Pull request by soundsonacid**

### CODE REVIEW

File: src/usermap.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        let usermap = self.usermap.clone();
+        let latest_slot = self.latest_slot.clone();
+
+        self.subscription
+            .event_emitter
+            .subscribe("usermap", move |event| {
+                if let Some(update) = event.as_any().downcast_ref::<ProgramAccountUpdate<User>>() {
+                    let user_data_and_slot = update.data_and_slot.clone();
+                    let user_pubkey = update.pubkey.to_string();
+                    if update.data_and_slot.slot > latest_slot.load(Ordering::Relaxed) {
```
Comment (jordy25519) [inline src/usermap.rs:85]: maybe `>=` ? so updates from different user in the same slot should both go through

**Review by jordy25519 — COMMENTED** [review]
would be good to add doc comments to the methods

File: src/usermap.rs (hunk 1 above)
Comment (soundsonacid) [reply src/usermap.rs:85]: the data updates will still go through, it just won't go to the effort of updating the latest slot if it's the same slot

File: src/usermap.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+use solana_client::rpc_config::{RpcAccountInfoConfig, RpcProgramAccountsConfig};
+use solana_client::rpc_request::RpcRequest;
+use solana_client::rpc_response::{OptionalContext, RpcKeyedAccount};
+use solana_sdk::commitment_config::CommitmentConfig;
+use solana_sdk::pubkey::Pubkey;
+
+pub struct Usermap {
+    subscribed: bool,
+    subscription: WebsocketProgramAccountSubscriber,
+    usermap: Arc<DashMap<String, User>>,
```
[the remark this answers is not in the document]
Comment (soundsonacid) [reply src/usermap.rs:27]: the pubkeys come out of the events as strings, i think keeping it string makes sense

File: src/usermap.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    pub fn size(&self) -> usize {
+        self.usermap.len()
+    }
+
+    pub fn contains(&self, pubkey: &str) -> bool {
+        self.usermap.contains_key(pubkey)
+    }
+
+    pub fn get(&self, pubkey: &str) -> Option<User> {
+        self.usermap.get(pubkey).map(|user| user.value().clone())
```
[the remark this answers is not in the document]
Comment (soundsonacid) [reply src/usermap.rs:114]: without using Cow i don't think so and i'd really rather not use Cow, cloning 4.3kb shouldn't be that expensive

**Pull request merged**